File: server/handler/public_ip.go
package handler

import (
	"fmt"
	"net/http"

	"github.com/gin-gonic/gin"

	"kvm_console/model"
	"kvm_console/service"
	"kvm_console/taskqueue"
)

func ListPublicIPs(c *gin.Context) {
	items, err := service.ListPublicIPs()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"code": 500, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"code": 200, "message": "ok", "data": items})
}

func CreatePublicIP(c *gin.Context) {
	var req service.PublicIPRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "参数错误"})
		return
	}
	row, err := service.CreatePublicIP(req)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"code": 200, "message": "公网 IP 已添加", "data": row})
}

// BatchCreatePublicIPs 批量新增公网 IP（管理员）
func BatchCreatePublicIPs(c *gin.Context) {
	var req service.PublicIPBatchRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "参数错误"})
		return
	}
	result, err := service.BatchCreatePublicIPs(req)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"code":    200,
		"message": fmt.Sprintf("成功新增 %d 个公网 IP（跳过 %d，失败 %d）", result.Created, result.Skipped, result.Failed),
		"data":    result,
	})
}

func DiscoverPublicIPv6Prefixes(c *gin.Context) {
	items, err := service.DiscoverPublicIPv6Prefixes(c.Query("uplink_if"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"code": 200, "message": "ok", "data": items})
}

func ImportPublicIPv6Prefix(c *gin.Context) {
	var req service.PublicIPv6PrefixImportRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "参数错误"})
		return
	}
	result, err := service.ImportPublicIPv6Prefix(req)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"code": 200, "message": "IPv6 前缀已导入", "data": result})
}

func UpdatePublicIP(c *gin.Context) {
	id, err := service.ParsePublicIPID(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	var req service.PublicIPRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "参数错误"})
		return
	}
	row, err := service.UpdatePublicIP(id, req)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"code": 200, "message": "公网 IP 已更新", "data": row})
}

func DeletePublicIP(c *gin.Context) {
	if !requireHighRiskVerification(c, "delete_public_ip") {
		return
	}
	id, err := service.ParsePublicIPID(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	if err := service.DeletePublicIP(id); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"code": 200, "message": "公网 IP 已删除"})
}

// BatchDeletePublicIPs 批量删除公网 IP（管理员，高风险）。
// 已绑定的 IP 自动跳过，部分失败不影响其他 IP。
func BatchDeletePublicIPs(c *gin.Context) {
	if !requireHighRiskVerification(c, "delete_public_ip") {
		return
	}
	var req service.PublicIPBatchOpRequest
	if err := c.ShouldBindJSON(&req); err != nil || len(req.IDs) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "参数错误: 需要公网 IP ID 列表"})
		return
	}
	result, err := service.BatchDeletePublicIPs(req.IDs)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"code":    200,
		"message": fmt.Sprintf("批量删除完成（成功 %d / 失败 %d / 跳过 %d）", result.Success, result.Failed, result.Skipped),
		"data":    result,
	})
}

// BatchBindPublicIPs 批量绑定公网 IP（管理员，高风险）。
// 通过任务队列异步应用规则，只提交一个任务，逐条绑定后统一应用。
func BatchBindPublicIPs(c *gin.Context) {
	if !requireHighRiskVerification(c, "bind_public_ip") {
		return
	}
	var req service.PublicIPBatchOpRequest
	if err := c.ShouldBindJSON(&req); err != nil || len(req.Items) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "参数错误: 需要批量绑定条目"})
		return
	}
	submitPublicIPTask(c, service.PublicIPOperationParams{Action: "batch_bind", BatchItems: req.Items})
}

// BatchUnbindPublicIPs 批量解绑公网 IP（管理员，高风险）。
// 通过任务队列异步应用规则，只提交一个任务，逐条解绑后统一应用。
func BatchUnbindPublicIPs(c *gin.Context) {
	if !requireHighRiskVerification(c, "unbind_public_ip") {
		return
	}
	var req service.PublicIPBatchOpRequest
	if err := c.ShouldBindJSON(&req); err != nil || len(req.IDs) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "参数错误: 需要公网 IP ID 列表"})
		return
	}
	items := make([]service.PublicIPBatchOpItem, 0, len(req.IDs))
	for _, id := range req.IDs {
		items = append(items, service.PublicIPBatchOpItem{PublicIPID: id})
	}
	submitPublicIPTask(c, service.PublicIPOperationParams{Action: "batch_unbind", BatchItems: items})
}

func PreviewPublicIP(c *gin.Context) {
	id, err := service.ParsePublicIPID(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	var req service.PublicIPBindRequest
	_ = c.ShouldBindJSON(&req)
	preview, err := service.PreviewPublicIPBinding(id, req)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"code": 200, "message": "ok", "data": preview})
}

func BindPublicIP(c *gin.Context) {
	if !requireHighRiskVerification(c, "bind_public_ip") {
		return
	}
	id, err := service.ParsePublicIPID(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	var req service.PublicIPBindRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "参数错误"})
		return
	}
	submitPublicIPTask(c, service.PublicIPOperationParams{Action: "bind", PublicIPID: id, BindRequest: req})
}

func UnbindPublicIP(c *gin.Context) {
	if !requireHighRiskVerification(c, "unbind_public_ip") {
		return
	}
	id, err := service.ParsePublicIPID(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	submitPublicIPTask(c, service.PublicIPOperationParams{Action: "unbind", PublicIPID: id})
}

func MigratePublicIP(c *gin.Context) {
	if !requireHighRiskVerification(c, "migrate_public_ip") {
		return
	}
	id, err := service.ParsePublicIPID(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	var req service.PublicIPBindRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "参数错误"})
		return
	}
	submitPublicIPTask(c, service.PublicIPOperationParams{Action: "migrate", PublicIPID: id, BindRequest: req})
}

func ApplyPublicIPRules(c *gin.Context) {
	if !requireHighRiskVerification(c, "apply_public_ip") {
		return
	}
	submitPublicIPTask(c, service.PublicIPOperationParams{Action: "apply_all"})
}

func submitPublicIPTask(c *gin.Context, params service.PublicIPOperationParams) {
	username, _ := c.Get("username")
	createdBy, _ := username.(string)
	task, err := taskqueue.SubmitWithStruct(model.TaskTypePublicIPApply, params, createdBy)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"code": 500, "message": "提交公网 IP 任务失败: " + err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"code":    200,
		"message": "公网 IP 任务已提交",
		"data": gin.H{
			"task_id": task.ID,
			"status":  task.Status,
		},
	})
}

File: web/src/api/publicIp.ts
/**
 * 公网 IP 相关 API（仅管理员）
 * 对应后端 /api/network/public-ips 路由组
 * 绑定/解绑/迁移/重载为高风险操作（428 二次验证由请求层自动处理），
 * 且通过任务队列异步应用规则
 */
import service from './client'
import type { ApiResponse } from '@/types/api'

/** 公网 IP 绑定模式 */
export type PublicIpMode = 'nat' | 'classic_route' | 'classic_bridge'

/** 公网 IP 绑定信息 */
export interface PublicIpBinding {
  id: number
  public_ip_id: number
  public_ip: string
  username: string
  vm_name: string
  vm_private_ip?: string
  mode: PublicIpMode
  runtime_status?: string
  guest_ipv6_status?: string
  guest_ipv6_message?: string
  config_hint?: string
  last_applied_at?: string
}

/** 公网 IP 列表项 */
export interface PublicIpItem {
  id: number
  ip: string
  cidr?: string
  gateway?: string
  uplink_if?: string
  supported_modes?: string
  status?: string // free / bound / reserved
  remark?: string
  modes?: PublicIpMode[]
  mode_labels?: string[]
  binding?: PublicIpBinding
  runtime_rules?: string[]
  issues?: string[]
  address_family?: 'ipv4' | 'ipv6'
  auto_ipv6?: boolean
  ipv6_source_prefix?: string
}

export interface PublicIPv6PrefixInfo {
  uplink_if: string
  address: string
  prefix: string
  gateway?: string
}

export interface PublicIPv6PrefixImportPayload {
  uplink_if: string
  prefix: string
  count: number
  remark?: string
}

export interface PublicIPv6PrefixImportResult {
  prefix: string
  created: PublicIpItem[]
  skipped: number
}

/** 公网 IP 创建/编辑请求 */
export interface PublicIpPayload {
  ip: string
  cidr?: string
  gateway?: string
  uplink_if?: string
  /** 支持模式，逗号分隔，如 "nat,classic_route" */
  supported_modes?: string
  status?: string // free / reserved
  remark?: string
}

/** 批量新增公网 IP 请求，共用除 IP 外的字段 */
export interface PublicIpBatchPayload {
  ips: string[]
  cidr?: string
  gateway?: string
  uplink_if?: string
  supported_modes?: string
  status?: string
  remark?: string
}

/** 批量新增中单条 IP 的处理结果 */
export interface PublicIpBatchItemResult {
  ip: string
  status: 'created' | 'skipped' | 'failed'
  reason?: string
  row?: PublicIpItem
}

/** 批量新增公网 IP 的汇总结果 */
export interface PublicIpBatchResult {
  created: number
  skipped: number
  failed: number
  items: PublicIpBatchItemResult[]
}

/** 绑定/迁移请求 */
export interface PublicIpBindPayload {
  username?: string
  vm_name?: string
  vm_private_ip?: string
  mode?: PublicIpMode
}

/** 规则预览结果 */
export interface PublicIpPreview {
  public_ip?: PublicIpItem
  binding?: PublicIpBindPayload
  commands?: string[]
  config_hint?: string
  warnings?: string[]
}

/** 任务提交响应 */
export interface PublicIpTaskResult {
  task_id?: string
  status?: string
}

/** 获取公网 IP 列表 */
export function getPublicIPs() {
  return service.get<unknown, ApiResponse<PublicIpItem[]>>('/network/public-ips', {
    silent: true,
  })
}

/** 新增公网 IP */
export function createPublicIP(data: PublicIpPayload) {
  return service.post<unknown, ApiResponse<PublicIpItem>>('/network/public-ips', data)
}

/** 批量新增公网 IP（共用除 IP 外的字段，一行一个） */
export function batchCreatePublicIPs(data: PublicIpBatchPayload) {
  return service.post<unknown, ApiResponse<PublicIpBatchResult>>(
    '/network/public-ips/batch',
    data,
  )
}

/** 检测上联网卡当前通过 RA 获取的公网 IPv6 前缀 */
export function discoverPublicIPv6Prefixes(uplinkIf?: string) {
  return service.get<unknown, ApiResponse<PublicIPv6PrefixInfo[]>>(
    '/network/public-ips/ipv6-prefixes',
    { params: uplinkIf ? { uplink_if: uplinkIf } : undefined },
  )
}

/** 将 IPv6 前缀展开为可逐台 VM 绑定的 /128 地址资源 */
export function importPublicIPv6Prefix(data: PublicIPv6PrefixImportPayload) {
  return service.post<unknown, ApiResponse<PublicIPv6PrefixImportResult>>(
    '/network/public-ips/ipv6-prefixes/import',
    data,
  )
}

/** 更新公网 IP */
export function updatePublicIP(id: number, data: PublicIpPayload) {
  return service.put<unknown, ApiResponse<PublicIpItem>>(`/network/public-ips/${id}`, data)
}

/** 删除公网 IP（高风险） */
export function deletePublicIP(id: number) {
  return service.delete<unknown, ApiResponse<unknown>>(`/network/public-ips/${id}`)
}

/** 预览绑定规则（试算） */
export function previewPublicIP(id: number, data: PublicIpBindPayload) {
  return service.post<unknown, ApiResponse<PublicIpPreview>>(
    `/network/public-ips/${id}/preview`,
    data,
  )
}

/** 绑定公网 IP（高风险，任务队列） */
export function bindPublicIP(id: number, data: PublicIpBindPayload) {
  return service.post<unknown, ApiResponse<PublicIpTaskResult>>(
    `/network/public-ips/${id}/bind`,
    data,
  )
}

/** 解绑公网 IP（高风险，任务队列） */
export function unbindPublicIP(id: number) {
  return service.post<unknown, ApiResponse<PublicIpTaskResult>>(
    `/network/public-ips/${id}/unbind`,
  )
}

/** 迁移公网 IP（高风险，任务队列） */
export function migratePublicIP(id: number, data: PublicIpBindPayload) {
  return service.post<unknown, ApiResponse<PublicIpTaskResult>>(
    `/network/public-ips/${id}/migrate`,
    data,
  )
}

/** 重载全部公网 IP 规则（高风险，任务队列） */
export function applyPublicIPRules() {
  return service.post<unknown, ApiResponse<PublicIpTaskResult>>('/network/public-ips/apply')
}

/** 批量操作单条结果 */
export interface PublicIpBatchOpResult {
  id: number
  ip: string
  status: 'success' | 'failed' | 'skipped'
  reason?: string
}

/** 批量操作汇总 */
export interface PublicIpBatchOpSummary {
  success: number
  failed: number
  skipped: number
  items: PublicIpBatchOpResult[]
}

/** 批量绑定单条参数 */
export interface PublicIpBatchBindItem {
  id: number
  payload: PublicIpBindPayload
}

/** 批量删除公网 IP（高风险，已绑定的自动跳过） */
export function batchDeletePublicIPs(ids: number[]) {
  return service.delete<unknown, ApiResponse<PublicIpBatchOpSummary>>('/network/public-ips/batch', {
    data: { ids },
  })
}

/** 批量解绑公网 IP（高风险，任务队列） */
export function batchUnbindPublicIPs(ids: number[]) {
  return service.post<unknown, ApiResponse<PublicIpTaskResult>>(
    '/network/public-ips/batch/unbind',
    { ids },
  )
}

/** 批量绑定公网 IP（高风险，任务队列） */
export function batchBindPublicIPs(items: PublicIpBatchBindItem[]) {
  return service.post<unknown, ApiResponse<PublicIpTaskResult>>(
    '/network/public-ips/batch/bind',
    {
      items: items.map((item) => ({
        public_ip_id: item.id,
        bind_request: item.payload,
      })),
    },
  )
}
